pkg/revisions/default_watcher.go

```diff
@@ -0,0 +1,126 @@
+// Copyright Istio Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package revisions
+
+import (
+	"log"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/api/label"
+	"istio.io/istio/pkg/kube"
+)
+
+const (
+	defaultTagWebhookName = "istio-revision-tag-default"
+)
+
+// DefaultWatcher keeps track of the current default revision and can notify watchers
+// when the default revision changes.
+type DefaultWatcher interface {
+	GetDefault() string
+	AddHandler(handler DefaultHandler)
+}
+
+type DefaultHandler func(string)
+
+type defaultWatcher struct {
+	revision        string
+	defaultRevision string
+	handlers        []DefaultHandler
+	webhookInformer cache.SharedInformer
+	mu              sync.Mutex
+}
+
+func NewDefaultWatcher(client kube.Client, revision string, stop <-chan struct{}) DefaultWatcher {
+	p := &defaultWatcher{
+		revision: revision,
+		mu:       sync.Mutex{},
+	}
+	webhookInformer := client.KubeInformer().Admissionregistration().V1().MutatingWebhookConfigurations().Informer()
+	webhookInformer.AddEventHandler(p.makeHandler())
+	go webhookInformer.Run(stop)
```

Comment:
Do not use .Run(), instead client.RunAndWait() (which is already called so you shouldn't need to at all, unless you are running it after startup)

Reply:
Done.